compiler/include/symbol.h

```diff
@@ -361,6 +379,45 @@ class ArgSymbol : public LcnSymbol {
 *                                                                   *
 ********************************* | ********************************/
 
+class ShadowVarSymbol : public VarSymbol {
+public:
+  ShadowVarSymbol(ForallIntentTag iIntent, const char* iName, Expr* iSpec = NULL);
+
+  virtual void    verify();
+  virtual void    accept(AstVisitor* visitor);
+  DECLARE_SYMBOL_COPY(ShadowVarSymbol);
+
+  virtual void    replaceChild(BaseAST* oldAst, BaseAST* newAst);
+  virtual bool    isConstant()                              const;
+  virtual bool    isConstValWillNotChange()                 const;
+
+  const char* intentDescrString() const;
+  // The corresponding outer var or NULL if not applicable.
+  SymExpr* outerVarSE() const { return (SymExpr*)outerVarRep; }
+  bool     isReduce()   const { return intent == TFI_REDUCE;  }
+  Expr*    spec()       const;
+  void     removeSupportingReferences();
+
+  ForallIntentTag intent;
+
+  // This would be a SymExpr*, if not for checkIdInsideWithClause().
+  // See also: sv->outerVarSE() and outerVarSym(sv).
```

Comment:
Huh? outerVarSE casts it to a SymExpr. Is that invalid in some situation? Should it use toSymExpr? It seems to me that it should be a SymExpr.

Reply:
Yes, it definitely should be a SymExpr. Except in one spot in scopeResolve (that's for checkIdInsideWithClause()) I need it to be something an UnresolvedSymExpr. Meaning that without it I'd have to replace checkIdInsideWithClause() with something ugly. :(

Reply:
Right, so in that event it seems to me the comment should say:
```
// Either a SymExpr* or an UnresolvedSymExpr*
// After scope resolve completes, there should be no more UnresolvedSymExprs
```

A reasonable alternative strategy would be to store the UnresolvedSymExpr in a separate field from the SymExpr.